### PR TITLE
refactor: Rewrite MetaTransactionMixin as composable

### LINKED ISSUE #4383: Rewrite MetaTransactionMixin as composable
Now our implementation is relying heavily on mixins that ofload a lot of logic.

It would be nice to replace it with some future proof implementation.

If you need to know how does it work feel free to ping me.

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

## Context

- [x] Closes #4383
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by yangwao** [comment]
what's about code climate notice? 👀
otherwise looks good!

**Comment by Jarsen136** [comment]
> what's about code climate notice? 👀 otherwise looks good!

will check it

File: composables/useMetaTransaction.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          },
+          (res) => resolveStatus(res.status)
+        )
+      )
+    } catch (e) {
+      if (e instanceof Error) {
+        const isCancelled = e.message === 'Cancelled'
+        if (isCancelled) {
+          showNotification(
+            $i18n.t('general.tx.cancelled'),
```
Comment (vikiival) [inline composables/useMetaTransaction.ts:59]: Please move callback to separated functions

File: composables/useTransaction.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { ExtrinsicStatus } from '@polkadot/types/interfaces'
+
+export default function () {
```
Comment (vikiival) [inline composables/useTransaction.ts:3]: may Rename `useTransactionStatus` ?

**Review by vikiival — CHANGES_REQUESTED** [review]

File: composables/useMetaTransaction.ts (hunk 1 above)
Comment (Jarsen136) [reply composables/useMetaTransaction.ts]: done

**Comment by Jarsen136** [comment]
> what's about code climate notice? 👀 otherwise looks good!

I have done some deeper refactor. Because we have both mixin version and  composable version of MetaTx function, there would be some `Duplication issue`. I think it's OK. In the future, we would keep only composable version and remove the other one.

**Comment by vikiival** [comment]
Just wondering, have you tested that ? 

Maybe good option to try it with the `teleport-bridge` you made before

**Comment by Jarsen136** [comment]
> Just wondering, have you tested that ?
> 
> Maybe good option to try it with the `teleport-bridge` you made before


Good point! However, I would not try it with the `teleport-bridge` because it dose not use MetaTransactionMixin right now.

I try the composable version of `useMetaTransaction` with Asset-list. I have successfully tested many times to switch fee asset between ksm and bsx.


image

✅ refactor AssetList component to composable api patten
✅ test `useMetaTransaction` functionality

**Review by vikiival — APPROVED** [review]
Code looks lgtm, thanks for the testing!

**Comment by yangwao** [comment]
pay 80 usd 
let it fly

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $80 @ 29.06 USD/KSM ~ 2.753 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0xb8bc2b7fe676ccffe35baed7b04b261dd2e3df2735e8e65d78a08b607ffa950f](https://kusama.subscan.io/extrinsic/0xb8bc2b7fe676ccffe35baed7b04b261dd2e3df2735e8e65d78a08b607ffa950f)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**